File: dsa/topics/10_trees/bt/java/Bt_View_BottomViewOfBinaryTree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.TreeMap;

import com.algos.topics.models.TreeNode;

public class Bt_View_BottomViewOfBinaryTree {
    public static void main(String[] args) {
        Bt_View_BottomViewOfBinaryTree obj = new Bt_View_BottomViewOfBinaryTree();

        TreeNode root = new TreeNode(3);
		root.left = new TreeNode(9);
        root.right = new TreeNode(20);
		root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);
        System.out.println(obj.bottomView(root));
    }

    // Idea: Use column number logic. Do col-1 when you are left and col+1 when going right.
    // Time: O(n)
    // space: O(1)
    public List<Integer> bottomView(TreeNode root){
        Map<Integer, Integer> map = new TreeMap<>();
        Queue<Tuple2> q = new LinkedList<>();
        q.add(new Tuple2(root, 0));
        while(!q.isEmpty()){
            Tuple2 curr = q.remove();
            TreeNode node = curr.node;
            int col = curr.col;

            map.put(col, node.val);

            if(node.left !=  null)
                q.add(new Tuple2(node.left, col-1));
            if(node.right !=  null)
                q.add(new Tuple2(node.right, col+1));
        }
        return new ArrayList<>(map.values());
    }
}
class Tuple2{
    TreeNode node;
    int col;
    public Tuple2(TreeNode node, int col) {
        this.node = node;
        this.col = col;
    }
}

File: dsa/topics/10_trees/bt/python/bt_view__bottom_view_of_binary_tree.py
# Binary Tree + BFS
# Problem: Bottom View of Binary Tree
# Idea:
# 1. Track horizontal column index for each node.
# 2. For each column, keep the latest (bottom-most) node value encountered in BFS.
# 3. Return values in order of sorted columns.
# Time: O(n)
# Space: O(n)

from collections import deque, OrderedDict

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Tuple2:
    def __init__(self, node, col):
        self.node = node
        self.col = col

def bottom_view(root):
    if not root:
        return []

    column_map = {}
    q = deque([Tuple2(root, 0)])

    while q:
        curr = q.popleft()
        node, col = curr.node, curr.col
        # Overwrite value for the column; last value in BFS at this column will be bottom view
        column_map[col] = node.val

        if node.left:
            q.append(Tuple2(node.left, col - 1))
        if node.right:
            q.append(Tuple2(node.right, col + 1))

    return [column_map[col] for col in sorted(column_map.keys())]

# -------- Test --------
root = TreeNode(3)
root.left = TreeNode(9)
root.right = TreeNode(20)
root.right.left = TreeNode(15)
root.right.right = TreeNode(7)

print(bottom_view(root))  # Expected: [9, 3, 15, 20, 7]
